# Eighty columns, one codebook

`data/train.csv` has 81 columns: `Id`, 79 clues, and `SalePrice`. `data/test.csv` has the same clues and no price. `data/data_description.txt` is the codebook that sits beside the tables. A few of its spellings differ from the cells you will actually read. The file says `NAmes` for North Ames; the codebook line says `Names`. The file says `BedroomAbvGr` and `KitchenAbvGr`; the codebook headings say Bedroom and Kitchen. Trust the cell, and use the codebook for the meaning.

The clues fall into eight families.

**The lot.** `MSSubClass` is a dwelling-type code, not a count. `MSZoning` is the zoning class. `LotFrontage` is street frontage in feet. `LotArea` is the lot in square feet. `Street` and `Alley` say how the lot is reached. `LotShape`, `LandContour`, `Utilities`, `LotConfig`, and `LandSlope` describe the parcel.

**The place.** `Neighborhood` is the area inside Ames. `Condition1` and `Condition2` record roads, railroads, or a nearby park. `BldgType` and `HouseStyle` say whether the building is a single family, a townhouse, one story, or two.

**The grades and the years.** `OverallQual` and `OverallCond` are grades from 1 to 10. `YearBuilt` and `YearRemodAdd` are the construction year and the remodel year.

**The shell.** `RoofStyle`, `RoofMatl`, `Exterior1st`, `Exterior2nd`, `MasVnrType`, `MasVnrArea`, `ExterQual`, `ExterCond`, and `Foundation`.

**The basement.** `BsmtQual`, `BsmtCond`, `BsmtExposure`, `BsmtFinType1`, `BsmtFinSF1`, `BsmtFinType2`, `BsmtFinSF2`, `BsmtUnfSF`, and `TotalBsmtSF`.

**The rooms.** `Heating`, `HeatingQC`, `CentralAir`, `Electrical`, then `1stFlrSF`, `2ndFlrSF`, `LowQualFinSF`, and `GrLivArea`. Baths are `BsmtFullBath`, `BsmtHalfBath`, `FullBath`, and `HalfBath`. Then `BedroomAbvGr`, `KitchenAbvGr`, `KitchenQual`, `TotRmsAbvGrd`, and `Functional`. `Fireplaces` and `FireplaceQu` are the hearth.

**The garage, the porch, the pool.** `GarageType`, `GarageYrBlt`, `GarageFinish`, `GarageCars`, `GarageArea`, `GarageQual`, `GarageCond`, and `PavedDrive`. The outdoor pieces are `WoodDeckSF`, `OpenPorchSF`, `EnclosedPorch`, `3SsnPorch`, `ScreenPorch`, `PoolArea`, `PoolQC`, `Fence`, `MiscFeature`, and `MiscVal`.

**The sale.** `MoSold`, `YrSold`, `SaleType`, and `SaleCondition`. Then `SalePrice`, training rows only.

A blank in this file is the two letters `NA`, not an empty string. For some columns the codebook defines `NA` as "none of this feature." For others it is an unknown measurement. The next lessons separate those two.


In [2]:
# Locate the contest tables beside this notebook, or under the course root.
from pathlib import Path
import csv

def manifest_path(name):
    here = Path.cwd()
    for folder in [here, *here.parents]:
        nested = folder / "14-House Prices" / "data" / name
        direct = folder / "data" / name
        if nested.exists():
            return nested
        if direct.exists() and (folder / "data" / "data_description.txt").exists():
            return direct
    raise FileNotFoundError(name)

def read_manifest(name):
    with manifest_path(name).open(newline="", encoding="utf-8") as handle:
        return list(csv.DictReader(handle))

train = read_manifest("train.csv")
test = read_manifest("test.csv")
submission = read_manifest("sample_submission.csv")
print(len(train), len(test), len(submission))
print(train[0]["Id"], train[0]["SalePrice"])


1460 1459 1459
1 208500


In [3]:
# The column order is part of the file, and so is the codebook's North Ames spelling.
expected = [
    "Id", "MSSubClass", "MSZoning", "LotFrontage", "LotArea", "Street", "Alley",
    "LotShape", "LandContour", "Utilities", "LotConfig", "LandSlope", "Neighborhood",
    "Condition1", "Condition2", "BldgType", "HouseStyle", "OverallQual", "OverallCond",
    "YearBuilt", "YearRemodAdd", "RoofStyle", "RoofMatl", "Exterior1st", "Exterior2nd",
    "MasVnrType", "MasVnrArea", "ExterQual", "ExterCond", "Foundation", "BsmtQual",
    "BsmtCond", "BsmtExposure", "BsmtFinType1", "BsmtFinSF1", "BsmtFinType2",
    "BsmtFinSF2", "BsmtUnfSF", "TotalBsmtSF", "Heating", "HeatingQC", "CentralAir",
    "Electrical", "1stFlrSF", "2ndFlrSF", "LowQualFinSF", "GrLivArea", "BsmtFullBath",
    "BsmtHalfBath", "FullBath", "HalfBath", "BedroomAbvGr", "KitchenAbvGr",
    "KitchenQual", "TotRmsAbvGrd", "Functional", "Fireplaces", "FireplaceQu",
    "GarageType", "GarageYrBlt", "GarageFinish", "GarageCars", "GarageArea",
    "GarageQual", "GarageCond", "PavedDrive", "WoodDeckSF", "OpenPorchSF",
    "EnclosedPorch", "3SsnPorch", "ScreenPorch", "PoolArea", "PoolQC", "Fence",
    "MiscFeature", "MiscVal", "MoSold", "YrSold", "SaleType", "SaleCondition",
    "SalePrice",
]
print(len(train[0]), len(test[0]))
assert list(train[0]) == expected
assert list(test[0]) == expected[:-1]
names = {row["Neighborhood"] for row in train}
print("NAmes" in names, "Names" in names, len(names))
assert "NAmes" in names
assert "Names" not in names
assert len(names) == 25


81 80
True False 25


## Pitfall

`MSSubClass` is stored as a number and still is not a quantity. Code 20 means a one-story house from 1946 or later. Code 30 means a one-story house from 1945 or earlier. Code 60 means a two-story house from 1946 or later. Adding one to the code does not add one room.

## Summary

1460 rows with a price, 1459 rows without one, 79 clues, and a codebook that you read against the cells. The token in the cell is `NA`. The neighborhood string in the cell is `NAmes`.
